### PR TITLE
v4: Framework for configuring, initialising and freeing libraries

### PR DESCRIPTION
**Pull request by ndptech**
e.g. libldap where there are global options which should be set once even though the library may be used by more than one FreeRADIUS module

### CODE REVIEW

File: src/lib/server/lib.c [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+ */
+static int lib_init_call(fr_lib_t *lib)
+{
+	CONF_SECTION	*global_cs, *cs;
+
+	/*
+	 *	Find relevant config section
+	 */
+	global_cs = cf_section_find(cf_root(main_config->root_cs), "global", NULL);
+	if (!global_cs) {
```
Comment (alandekok) [inline src/lib/server/lib.c:54]: I'd just have this, and the `lib->name` check return 0 if there's no configuration section found.  "Do nothing" should be a perfectly fine configuration.

File: src/lib/server/lib.c [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+	for (; (*p)->name; p++) {
+		fr_lib_t	*lib = NULL;
+
+		lib = fr_rb_find(&lib_list->libs, &(fr_lib_t){ .name = (*p)->name });
+
+		/*
+		 *  If the library is already initialised, just increase the reference count
+		 */
+		if ((lib) && (lib->initialised)) {
```
Comment (alandekok) [inline src/lib/server/lib.c:100]: what about `if lib && !lib->initialized` ?  Maybe make that an assert?

Comment (ndptech) [reply src/lib/server/lib.c:100]: That's a valid thing which could happen.  When libraries are loaded during config parsing, the entry would be created in the tree on the first occasion the library is loaded, however initialisation can't happen until the config parsing is completed as `main_config->root_cs` is not populated until that point.
So if the same library is used by more than one module in FR which causes it to be loaded more than once during config parsing, after the first time it is loaded, subsequent calls will have `lib && !lib-initialised`.  The initialisation will happen after config parsing has been completed by a call to `lib_init()`

File: src/lib/server/lib.c (hunk 1 above)
Comment (arr2036) [reply src/lib/server/lib.c:54]: Agreed, warnings not required here.

File: src/lib/server/lib.c [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+	global_cs = cf_section_find(cf_root(main_config->root_cs), "global", NULL);
+	if (!global_cs) {
+		WARN("Missing global config section");
+		goto init;
+	}
+
+	cs = cf_section_find(global_cs, lib->name, NULL);
+	if (!cs) {
+		WARN("Missing \"%s\" subsection in global config", lib->name);
+		goto init;
```
Comment (arr2036) [inline src/lib/server/lib.c:62]: You still need to evaluate the default rules.  Probably easiest to alloc the sections if they're not found and just add them in.

File: src/lib/server/lib.c [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+	for (; (*p)->name; p++) {
+		fr_lib_t	*lib = NULL;
+
+		lib = fr_rb_find(&lib_list->libs, &(fr_lib_t){ .name = (*p)->name });
+
+		/*
+		 *  If the library is already initialised, just increase the reference count
+		 */
+		if ((lib) && (lib->initialised)) {
+			lib->instance_count ++;
```
Comment (arr2036) [inline src/lib/server/lib.c:102]: WS

File: src/lib/server/lib.c (hunk 2 above)
Comment (arr2036) [reply src/lib/server/lib.c:101]: Agree with @ndptech.  Multiple modules use things like `libfreeradius-curl`, and this is the global config, i.e. there's one config structure for the entire process.

File: src/lib/server/lib.c [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+/*
+ * $Id$
+ *
+ * @file lib.c
```
Comment (arr2036) [inline src/lib/server/lib.c:20]: This is a bit of a generic name, maybe `global.c` instead?

File: src/lib/server/lib.h [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+	fr_rb_node_t		entry;			//!<  Entry in tree of libraries
+	void			*data;			//!<  Library configuration data
+	fr_lib_autoinit_t	*autoinit;		//!<  Autoinit structure used to manage this library
+	uint32_t		instance_count;		//!<  Number of current uses of this library
+	bool			initialised;		//!<  Has the init callback been run for this library
+} fr_lib_t;
+
+typedef struct {
+	fr_rb_tree_t		libs;
+} fr_lib_list_t;
```
Comment (arr2036) [inline src/lib/server/lib.h:65]: Move into the .c file

File: src/lib/server/lib.h [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+ *
+ */
+typedef struct {
+	char const		*name;			//!<  Name of library
+	fr_rb_node_t		entry;			//!<  Entry in tree of libraries
+	void			*data;			//!<  Library configuration data
+	fr_lib_autoinit_t	*autoinit;		//!<  Autoinit structure used to manage this library
+	uint32_t		instance_count;		//!<  Number of current uses of this library
+	bool			initialised;		//!<  Has the init callback been run for this library
+} fr_lib_t;
```
Comment (arr2036) [inline src/lib/server/lib.h:61]: Move into the .c file

File: src/lib/server/lib.h [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+typedef struct {
+	fr_rb_tree_t		libs;
+} fr_lib_list_t;
+
+int fr_dl_lib_autoinit(dl_t const *module, void *symbol, void *user_ctx);
```
Comment (arr2036) [inline src/lib/server/lib.h:67]: Don't need the fr_ prefix that's only for functions in the util library and the protocol libraries, i.e. stuff that needs to be namespaced.  The server library is there so we can call the same code from the test harness binaries and `radiusd`, it's not meant for external consumption in the same way.

File: src/lib/server/lib.h [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+/** Structure to define how to initialise libraries with global configuration
+ *
+ */
+typedef struct {
+	char const 		*name;			//!<  Name of library and section within global config
+	CONF_PARSER const	*config;		//!<  Config parser for this library's global options
+	void			*inst;			//!<  Module data to parse global config into
+	fr_lib_init_t		init;			//!<  Callback to initialise library
+	fr_lib_free_t		free;			//!<  Callback to free library
+} fr_lib_autoinit_t;
```
Comment (arr2036) [inline src/lib/server/lib.h:44]: `global_lib_autoinit_t`?

File: src/lib/server/lib.c [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+
+/** Run free callbacks for external libraries no-longer in use
+ *
+ * @param[in] to_free	Array of autoinit structures detailing libraries to free
+ */
+static void lib_autofree(fr_lib_autoinit_t * const *to_free)
+{
+	fr_lib_autoinit_t * const *p = to_free;
+
+	for (; (*p)->name; p++) {
```
Comment (arr2036) [inline src/lib/server/lib.c:143]: Maybe stick the init on the same line.  Just makes it slightly clearer.

File: src/lib/server/lib.c [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+		fr_lib_t	*lib = NULL;
+
+		lib = fr_rb_find(&lib_list->libs, &(fr_lib_t){ .name = (*p)->name });
+		if (!lib) {
```
Comment (arr2036) [inline src/lib/server/lib.c:147]: I think this should probably be an assert, as there shouldn't be any way this happens normally.

File: src/lib/server/lib.c [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+ * @param to_init	Array of autoinit structures detailing libraries to initialise
+ * @return
+ *	- 0 on succcess
+ *	- -1 on failure
+ */
+static int lib_autoinit(fr_lib_autoinit_t * const *to_init)
+{
+	fr_lib_autoinit_t * const *p = to_init;
+
+	for (; (*p)->name; p++) {
```
Comment (arr2036) [inline src/lib/server/lib.c:93]: Or `p != fr_lib_autoinit_terminator`.  Ensures people actually use the terminator :)

File: src/lib/server/lib.c [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+ *
+ */
+static int8_t _lib_name_cmp(void const *one, void const *two)
+{
+	fr_lib_t const	*a = one;
+	fr_lib_t const	*b = two;
+	int 		ret;
+
+	ret = strcmp(a->name, b->name);
+	return CMP(ret, 0);
```
Comment (arr2036) [inline src/lib/server/lib.c:182]: `return CMP(a->autoinit, b->autoinit)`?

File: src/lib/server/lib.c (hunk 12 above)
Comment (arr2036) [reply src/lib/server/lib.c:93]: Maybe stick the init on the same line. Just makes it slightly clearer.

File: src/lib/server/lib.c (hunk 10 above)
Comment (arr2036) [reply src/lib/server/lib.c:143]: Or p != fr_lib_autoinit_terminator. Ensures people actually use the terminator :)

File: src/lib/server/lib.c (hunk 11 above)
Comment (arr2036) [reply src/lib/server/lib.c:147]: `.autoinit = p`

File: src/lib/server/lib.c [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+ *	- -1 on failure
+ */
+static int lib_autoinit(fr_lib_autoinit_t * const *to_init)
+{
+	fr_lib_autoinit_t * const *p = to_init;
+
+	for (; (*p)->name; p++) {
+		fr_lib_t	*lib = NULL;
+
+		lib = fr_rb_find(&lib_list->libs, &(fr_lib_t){ .name = (*p)->name });
```
Comment (arr2036) [inline src/lib/server/lib.c:96]: `.autoinit = p`

File: src/lib/server/lib.h [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+extern const fr_lib_autoinit_t fr_lib_autoinit_terminator;
+
+/** Structure to track use of libraries.
+ *
+ */
+typedef struct {
+	char const		*name;			//!<  Name of library
+	fr_rb_node_t		entry;			//!<  Entry in tree of libraries
+	void			*data;			//!<  Library configuration data
+	fr_lib_autoinit_t	*autoinit;		//!<  Autoinit structure used to manage this library
```
Comment (arr2036) [inline src/lib/server/lib.h:58]: `const`

File: src/lib/server/lib.h [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+/*
+ *	To be used as terminator in an array of fr_lib_autoinit_t
+ */
+extern const fr_lib_autoinit_t fr_lib_autoinit_terminator;
+
+/** Structure to track use of libraries.
+ *
+ */
+typedef struct {
+	char const		*name;			//!<  Name of library
```
Comment (arr2036) [inline src/lib/server/lib.h:55]: no longer needed

File: src/lib/server/lib.h [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+
+typedef int (*fr_lib_init_t)(void);
+
+typedef void (*fr_lib_free_t)(void);
+
+/** Structure to define how to initialise libraries with global configuration
+ *
+ */
+typedef struct {
+	char const 		*name;			//!<  Name of library and section within global config
```
Comment (arr2036) [inline src/lib/server/lib.h:39]: no longer needed

File: src/lib/server/lib.h [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+	CONF_PARSER const	*config;		//!<  Config parser for this library's global options
+	void			*inst;			//!<  Module data to parse global config into
```
[the remark this answers is not in the document]
Comment (arr2036) [reply src/lib/server/lib.h:41]: ...wait, what's the data field in `fr_lib_t`?

File: src/lib/server/lib.h [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+ */
+extern const fr_lib_autoinit_t fr_lib_autoinit_terminator;
+
+/** Structure to track use of libraries.
+ *
+ */
+typedef struct {
+	char const		*name;			//!<  Name of library
+	fr_rb_node_t		entry;			//!<  Entry in tree of libraries
+	void			*data;			//!<  Library configuration data
```
Comment (arr2036) [inline src/lib/server/lib.h:57]: Doesn't look like this is used?

File: src/lib/server/lib.c (hunk 10 above)
Comment (arr2036) [reply src/lib/server/lib.c:143]: Maybe stick the init on the same line. Just makes it slightly clearer.

File: src/lib/server/global_lib.c [hunk 20]
```diff
@@ -0,0 +1,232 @@
...
+ *  Terminator for array of global_lib_autoinit_t
+ */
+global_lib_autoinit_t const lib_autoinit_terminator = { .name = NULL };
+
+/*
+ *  Global list of libraries
+ */
+typedef struct {
+	fr_rb_tree_t		libs;
+} fr_lib_list_t;
```
Comment (arr2036) [inline src/lib/server/global_lib.c:39]: Take off the `fr_` prefix from anything in the server library.

File: src/lib/server/global_lib.h [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+	lib_init_t		init;			//!<  Callback to initialise library
+	lib_free_t		free;			//!<  Callback to free library
+} global_lib_autoinit_t;
+
+/*
+ *	To be used as terminator in an array of fr_lib_autoinit_t
+ */
+extern const global_lib_autoinit_t lib_autoinit_terminator;
+
+int dl_lib_autoinit(dl_t const *module, void *symbol, void *user_ctx);
```
Comment (arr2036) [inline src/lib/server/global_lib.h:51]: global_lib_autoinit

File: src/lib/server/global_lib.h [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+
+void dl_lib_autofree(dl_t const *module, void *symbol, void *user_ctx);
```
Comment (arr2036) [inline src/lib/server/global_lib.h:53]: global_lib_autofree

File: src/lib/server/global_lib.h [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+
+int lib_list_init(void);
+
+int lib_init(void);
```
Comment (arr2036) [inline src/lib/server/global_lib.h:57]: global_list_init

File: src/lib/server/global_lib.c (hunk 20 above)
Comment (arr2036) [reply src/lib/server/global_lib.c:39]: `global_list_list_t`

**Pull request merged**